dandalandone/POC-EyCommerce
# Done creating RBAC, CASL authorization

## Ticket/s

- [User Authorization](https://trello.com/c/IjnkaW52/3-api-user-user-authorization)

## Changes

- Added user data (user id, user type) when accessing the token.
- Created a role guard and role decorator.
- Installs @casl/ability for permission based authorization
- Created ability factory, CheckAbility Decorator and Ability Guard.

## Screenshot

- add GIF or screenshot here

## How to test?

- Add AuthModule and AbilityModule to the destination module
```
@Module({
  imports: [
    MongooseModule.forFeature([{ name: Product.name, schema: ProductSchema }]),
    MerchantsModule,
    AuthModule,
    AbilityModule,
  ],
```
- Add to InferSubject of the target schema to Subjects type.
```
export type Subjects = InferSubjects<typeof User | typeof Product> | 'all';
```
- Add and define abilities to ability factory.
```
can(Action.Create, User);
      can(Action.Read, User);
      can(Action.Update, User);
      can(Action.Delete, User);

      can(Action.Create, Product);
```

- Import UseGuard Decorator and add AuthGuard as an argument.
   - this will check if the route is accessible for public or not and if the user is authorized via token.
```
@UseGuards(AuthGuard)
```
- Import Role decorator and import Role enum (role.enum.ts) to define the role to check.
   - define the user type that can access this route
```
  @Roles(Role.CONSUMER )
```
- Import UseGuard and pass in the RolesGuard as the parameter.
  - this checks whether the user can access the route based on her pre-defined role
```
@UseGuards(RolesGuard)
```
- Import CheckAbilities decorator anddefine the action and subject.
```
  @CheckAbilities({ action: Action.Update, subject: Product })
```

- Import UseGuard and pass in the AbilityGuard as the parameter.
  - Checks whether the ability defined from the CheckAbility decorator is allowed.
 
```
  @UseGuards(AbilityGuard)
```


```
curl --location 'http://localhost:3001/api/user/64d058b031a1aed28f258d5c' \
--header 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6ImFkcmlhbmNoaW5qZW4iLCJzdWIiOiI2NGQwNThiMDMxYTFhZWQyOGYyNThkNWEiLCJpYXQiOjE2OTE1NDk1NTEsImV4cCI6MTY5MTYzNTk1MX0.x3MSy6Yv0jBgDs9bXSBjPXdwSSi8aXSRuFo2OegwuTs'
```
